File: Leetcode/24-Swap-Nodes-in-Pairs/SwapNodesInPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* A = head;
        head = A->next;
        ListNode* B;
        while (A!=NULL) {
            B = A->next;
            if (B == NULL) {
                break;
            }

            A->next = B->next;
            B->next = A;
            dummy->next = B;
            dummy = A;
            A = A->next;
        }
        return head;
    }
};

File: Leetcode/24-Swap-Nodes-in-Pairs/SwapNodesInPairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	dummy := &ListNode{Val: 0, Next: nil}
	dummy.Next = head
	A := head
	head = A.Next
	var B *ListNode
	for A != nil {
		B = A.Next
		if B == nil {
			break
		}
		A.Next = B.Next
		B.Next = A
		dummy.Next = B
		dummy = A
		A = A.Next
	}
	return head
}
